staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go

```diff
@@ -40,107 +106,163 @@ type uniformClient struct {
 	nCalls   int
 	// duration for a simulated synchronous call
 	execDuration time.Duration
-	// duration for simulated "other work"
+	// duration for simulated "other work".  This can be negative,
+	// causing a request to be launched a certain amount of time
+	// before the previous one finishes.
 	thinkDuration time.Duration
+	// When true indicates that only half the specified number of
+	// threads should run during the first half of the evaluation
+	// period
+	split bool
 }
 
 // exerciseQueueSetUniformScenario runs a scenario based on the given set of uniform clients.
 // Each uniform client specifies a number of threads, each of which alternates between thinking
 // and making a synchronous request through the QueueSet.
 // This function measures how much concurrency each client got, on average, over
-// the initial evalDuration and tests to see whether they all got about the same amount.
-// Each client needs to be demanding enough to use this amount, otherwise the fair result
-// is not equal amounts and the simple test in this function would not accurately test fairness.
-// expectPass indicates whether the QueueSet is expected to be fair.
+// the initial evalDuration and tests to see whether they all got about the fair amount.
+// Each client needs to be demanding enough to use more than its fair share,
+// or overall care needs to be taken about timing so that scheduling details
+// do not cause any client to actually request a significantly smaller share
+// than it theoretically should.
+// expectFair1 and expectFair2 indicate whether the QueueSet is expected to be
+// fair in the respective halves of a split scenario;
+// in a non-split scenario expectFair1 is what matters.
 // expectedAllRequests indicates whether all requests are expected to get dispatched.
-func exerciseQueueSetUniformScenario(t *testing.T, name string, qs fq.QueueSet, sc uniformScenario,
+func exerciseQueueSetUniformScenario(t *testing.T, name string, qs fq.QueueSet,
+	sc uniformScenario,
+	concurrencyLimit int,
 	evalDuration time.Duration,
-	expectPass, expectedAllRequests, expectInqueueMetrics, expectExecutingMetrics bool,
+	expectFair1, expectFair2, expectedAllRequests bool,
+	evalInqueueMetrics, evalExecutingMetrics bool,
```

Comment:
This is a lot of parameters and this function is enormous, it might make sense to rephrase this code as an "options" struct with a number of methods. If not in this PR then can we agree to make the next person to touch this, fix it? :) (and add a TODO)

Reply:
I'm still reading this function but it is really hard to follow.

Reply:
I think I can fix both problems together.  Maybe later today.

Reply:
factored